### PR TITLE
added UI error handling

### PR DESCRIPTION
**Pull request by sakekasi**

### CODE REVIEW

File: backend/frontend/src/utils/webapi.js [hunk 1]
```diff
@@ -206,9 +226,9 @@ export function init(callback: any){
     }, handleError("init: Error requesting roots:"))
   .then((data) => {
       piazzaClassId = data.piazza_cid;
-      setTimeout(function(){
+      // setTimeout(function(){
```
Comment (nfischer) [inline backend/frontend/src/utils/webapi.js]: Why did we get rid of the timeout?

**Comment by sakekasi** [comment]
it was getting in the way of error handling. I tried it out, and it wasn't crashing anymore

File: backend/frontend/src/utils/webapi.js [hunk 2]
```diff
@@ -114,8 +115,12 @@ var posts = new Map();
 var cached = null;
 function getPiazzaPosts(courseId: number){
   let endpoint = mainUrl + `/${courseId}/course/getpiazzaposts/`;
-  http.get(endpoint, function(res){
+  let ignoredata = false;
+  var req = http.get(endpoint, function(res){
     res.on('data', function(buf){
+      if(ignoredata){
```
Comment (nfischer) [inline backend/frontend/src/utils/webapi.js:117]: Does `ignoreData` signify if we have or don't have a valid piazza ID? If not, what is the purpose of it, at a high level?

Comment (sakekasi) [inline backend/frontend/src/utils/webapi.js:117]: It specifies that we have encountered some kind of error, and that further data should be skipped

File: backend/frontend/gulpfile.js [hunk 3]
```diff
@@ -53,7 +53,7 @@ function build_node(){
 
   return gulp.src("src/**/*.js")
       .pipe(sourcemaps.init())
-      .pipe(babel({ optional: ['runtime'] }))
+      // .pipe(babel({ optional: ['runtime'] }))
```
Comment (nfischer) [inline backend/frontend/gulpfile.js]: Why did we remove babel?

**Comment by nfischer** [comment]
Overall, looks good. I will test later tonight or tomorrow, and give more thorough feedback once I do so.

File: backend/frontend/gulpfile.js (hunk 3 above)
Comment (sakekasi) [inline backend/frontend/gulpfile.js]: oh jeez, didn't notice I did that. Those two build targets are irrelevant tho. they are never used, and were just part of the gulpfile i copy-pasted from an old project

Comment (nfischer) [inline backend/frontend/gulpfile.js]: If we're not using them, let's just delete the lines if that makes sense. Also, will this remove babel from our package dependencies? Any dependencies we can safely remove, we probably should do so.

File: backend/frontend/src/utils/webapi.js (hunk 2 above)
Comment (nfischer) [inline backend/frontend/src/utils/webapi.js:117]: What happens if someone leaves a piazza ID blank for a course? Does that mean you'll never do the API call to begin with?

**Comment by nfischer** [comment]
Please allow support for empty piazza IDs. This is for courses that don't have an associated piazza forum. Additionally, if a piazza ID for a course is blank, the frontend should either:

1) not call `getpiazzaposts` (preferable), or
2) call `getpiazzaposts` but expect a failure

One way to implement option 1 would be to request the piazza ID before calling `getpiazzaposts`. Piazza ID will either be null or it will be the empty string, depending on the frontend's implementation during course creation. Either way, it's easy to check for.

Option 1 is preferable, since it means there's no dependency on saving piazza credentials server side (they'd never be checked if we never call `getpiazzaposts`).

File: backend/frontend/src/utils/webapi.js (hunk 2 above)
Comment (sakekasi) [inline backend/frontend/src/utils/webapi.js:117]: I think we do the call, but we get an error at this point

**Comment by nfischer** [comment]
Seems to behave pretty nicely. Looks good to merge

**Pull request merged**